sccp/decomposition.py

```diff
@@ -0,0 +1,105 @@
+import os
+import numpy as np
+from os.path import join
+import numpy as np
+from .parafac2 import parafac2_nd
+from tensorly.decomposition._parafac2 import _parafac2_reconstruction_error
+from tensorly.tenalg import khatri_rao
+from tensorpack import Decomposition
+from sklearn.decomposition import NMF
+
+path_here = os.path.dirname(os.path.dirname(__file__))
+
+def plotR2X(tensor, rank, datatype, ax, run_decomp=False, Inlclude_NNMF=False):
+    """Creates R2X plot for parafac2 tensor decomposition"""
+    rank_vec = np.arange(1, rank + 1)
+    if run_decomp == True:
+        pf2_error = np.empty(len(rank_vec))
+        pca_error = pf2_error.copy()
+        nnmf_error = pf2_error.copy()
+        for i in range(len(rank_vec)):
+            weights, factors, projs = parafac2_nd(
+                    tensor,
+                    rank=i+1,
+                    verbose=True
+                )
+
+            if len(tensor) > 3:
+                projs = np.reshape(projs, (-1, tensor.shape[-2], i+1))
+                factors = [khatri_rao(factors[:-2]), factors[-2], factors[-1]]
+
+            pf2_error[i] = 1 - _parafac2_reconstruction_error(
+                tensor, (weights, factors, projs))/ np.linalg.norm(tensor) ** 2
+            
+            decomp = Decomposition(tensor, max_rr=i+1)
```

Comment:
This gives you the reconstruction error with PCA for all ranks, so this can be outside the loop.